daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/infoset/InfosetImpl.scala

```diff
@@ -1133,7 +1133,14 @@ sealed trait DIElement
    * know if it is nil or not (i.e., hasn't be set, nor has
    * anything been set to indicate that it won't be nilled.)
    */
-  def isNilled: Boolean
+  def isNilled: Boolean = {
+    if (!erd.isNillable) false
+    else if (_isNilledSet) {
+      _isNilled
+    } else {
+      erd.toss(InfosetNoDataException(this, erd))
```

Comment:
If this is uncovered, is it even possible? If so we should create a test that drives it. If not it should be an assertion not a throw/toss.

Reply:
I've updated this PR to use the above logic. I think it simplfies things quite a bit. Only question is if there is a case I haven't thought of that would allow access to isNilled state before the nil parser completes.

Reply:
I have a vague memory of isNilledSet being needed for unparser close/open logic of nillable complex elements? ... maybe while debugging? 

Something about being able to answer fn:exists(...) exipressions about children of complex types without suspending maybe? or being able to ask if it has be set nil or not bypassing an error check. 

The debugging may be unrelated, except that when debugging, you may be paused at a step and want to eval and expression, and a complex type that is nilled, you want evaluation to not hang on fn:exists of its children waiting for some close event. 

There must be calls to isNilledSet to analyze shouldn't there?

Reply:
From what I can tell, unparsing shouldn't have any issues with this change. This is where we create infoset elements from InfosetInputters:

https://github.com/apache/daffodil/blob/564424403ce578b16d993507ccaa8da0a7d7d78e/daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/infoset/InfosetInputter.scala#L413-L430

And one of the first things we do there is get the nilled state from the InfosetInputter and set it on the infoset. So after that point, an infoset used for unparsing always has the correct nilled state and can always answer things like fn:exists/fn:nilled accurately and immediately, at least as far as the nilled state goes.


There were no direct calls to isNilledSet, it would just throw an exception if you asked isNilled and it wasn't set. But there were a few places that called maybeIsNilled which worked around this and would return None if isNilledSet was false. But I think that was needed because we never (or pretty late) set `_isNilledSet = true` for non-nilled elements. Some places wanted to know before that was set, and if it wasn't set they (correctly) assumed false. From what I can tell, the places that did this all happened after a NilParser has already figured isNilled out and set it appropriately.

My most recent commit changes it so we set nilled state asap, even for non-nilled elements, so those cases can safely use isNilled a know that isNilled is correct.

Reply:
I'm willing to concede that this may not be needed. I'm pretty sure this was about unparsing though, not parsing, and expressions that want to know fn:exists or fn:exactlyOne. I.e.., tests for existence of children in the context of suspensions.

There *may* have been some thought that the Infoset API could be used programatically to construct infosets and so the orderly behavior of infoset inputters could not be depended upon. We don't support that officially, so I'm fine with this change if all tests pass.

Reply:
In addition to all Daffodil tests passing, I also ran this change against the regression suite. There are a few unrelated test failures (changes in validation messages due to validationMode=full not doing Daffodil validation, hiddenGroupRefs no longer allowed on a complex type model group), but no failures that look related to this change. So I *think* this is probably safe to merge. If we do find an edge case where we do need to know if nilled is set, we can always revert this change or see if there's another workaround, but I think this does simplify things enough to make it worth it.